engine/src/gfx/tvector.h

```diff
@@ -0,0 +1,180 @@
+#ifndef TVECTOR_H
+#define TVECTOR_H
+
+#ifdef __cplusplus
+
+//#include "vectorizable.h"
+
+class Matrix;
+
+/*
+ * XVector = Vector, QFLOAT = float, YVector = QVector
+*/
+
+
+// Moved to start to prevent 'specialization ... after instantiation' error
+template<typename T>
+T _netswap(T);
+
+template<typename T>
+T _sqrt(T);
+
+
+
+// Note that S is QFLOAT and the main typename. T is used for defining YVector.
+template <typename S, typename T>
+class TVector {
+// Fields
+public:
+    union
+    {
+        S i;
+        S x;
+    };
+    union
+    {
+        S j;
+        S y;
+    };
+    union
+    {
+        S k;
+        S z;
+    };
+
+// Constructors
+    TVector() : i(0),j(0),k(0){}
+    TVector( const TVector<T,S> &a );
```

Comment:
Question: Why is this `T,S` and not `S,T`?

I have the same question about a few other lines below, also.

Reply:
@royfalk ?

Reply:
Quick disclaimer - still a beginner C++ dev, as I learned today to my sorrow. This is not a default copy constructor. That one is covered by the rule of 0. This is a constructor to create QVector from Vector and vice-versa. I worked off the original xvector.h and vec.h. If you look there, you'll find my reference constructors.

Reply:
@royfalk Thanks for the reply.

Perhaps the default copy constructor is implied. But based on my logging, the program was hitting this one quite a lot, until I defined the default copy constructor `<S,T>` explicitly, alongside this one.

Will it hurt anything to do this?

Reply:
@stephengtuggy that might mean we could simply do:
```
typedef TVector<float, double> Vector
typedef Vector QVector
```

Not sure, but it shouldn't hurt anything. Might help if we put a high level verbose logs in each so we can trace which is getting called too (e.g logs we have to explicitly enable for dev/diag purposes but would normally be turned off - probably a compile-time option).

Reply:
@BenjamenMeyer Hmm. Not sure. Wouldn't that mean that both `Vector` and `QVector` would be `float` precision, rather than `double`?

I do have the logging set up much the way you described. It's at the highest verbosity level -- `trace`, in Boost.Log terms -- which only gets enabled when you pass `--debug=3` to vegastrike on the command line. (As opposed to `--debug=2`, `--debug=1`, or `--debug` by itself.)

This whole thing is turning out to be quite the rabbit hole. I think I could use more eyes on it.

Reply:
@stephentuggy yeah...not sure. There's certainly some complications. Honestly we should push it to be all doubles and convert to float only when necessary.

Reply:
@BenjamenMeyer yeah, that was my thought as well. Ideally it would be all doubles. However, there are explicit `float`s all over the code. It would be a massive change.

Reply:
@stephentuggy. Do we want to wait until 0.9.x to do that?